File: HTTPShortcuts/app/src/main/java/ch/rmy/android/http_shortcuts/adapters/BaseViewHolder.java
package ch.rmy.android.http_shortcuts.adapters;

import android.support.v7.widget.RecyclerView;
import android.view.View;

import ch.rmy.android.http_shortcuts.realm.models.HasId;
import io.realm.RealmObject;

abstract class BaseViewHolder<T extends RealmObject & HasId> extends RecyclerView.ViewHolder {

    private T item;

    BaseViewHolder(View parent, final BaseAdapter<T> baseAdapter) {
        super(parent);
        itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (baseAdapter.clickListener != null) {
                    baseAdapter.clickListener.onItemClicked(item);
                }
            }
        });
        itemView.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                if (baseAdapter.clickListener != null) {
                    baseAdapter.clickListener.onItemLongClicked(item);
                    return true;
                }
                return false;
            }
        });
    }

    public void setItem(T item) {
        this.item = item;
        updateViews(item);
    }

    protected abstract void updateViews(T item);

}

File: HTTPShortcuts/app/src/main/java/ch/rmy/android/http_shortcuts/adapters/BaseViewHolder.kt
package ch.rmy.android.http_shortcuts.adapters

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import ch.rmy.android.http_shortcuts.realm.models.HasId
import io.realm.RealmObject

abstract class BaseViewHolder<in T>(parent: ViewGroup, layoutRes: Int, baseAdapter: BaseAdapter<T>) : RecyclerView.ViewHolder(LayoutInflater.from(parent.context).inflate(layoutRes, parent, false)) where T : RealmObject, T : HasId {

    private var item: T? = null

    init {
        itemView.setOnClickListener {
            if (item?.isValid == true) {
                baseAdapter.clickListener?.invoke(item!!)
            }
        }
        itemView.setOnLongClickListener {
            if (item?.isValid == true) {
                baseAdapter.longClickListener?.invoke(item!!) == true
            } else {
                false
            }
        }
    }

    fun setItem(item: T) {
        this.item = item
        updateViews(item)
    }

    protected abstract fun updateViews(item: T)

}